### PR TITLE
WIP: Feature/buildcache

### LINKED ISSUE #111: design cache mechanism
## Issue Description
sealer build will generate tmp dir for immediate layers, but never cache them, which result in repeating same operations at every build.
Type: *feature request*

### Describe what feature you want
design and finish the cache ability, then the issue [#62](https://github.com/alibaba/sealer/pull/62) of image id could be resolved

### PR DESCRIPTION
**Pull request by touchuyht**
### Describe what this PR does / why we need it
Build cache mechanism enables sealer to use the previous built image layers to increase image build speed.

### Does this pull request fix one issue?

Fixes #111 
### Describe how you did it
**imageID**: ImageID is calculated from the HASH256 hash of its discription yaml file.
**layerID**: Layers are referenced by cryptographic hashes of their serialized representation. This is a SHA256 digest over the tar archive used to transport the layer.

**StoreID**: StoreID is caculated from a series of serialized layers. The layers hash value is "", but the layers value and type is not empty. It is used in image store to help to reference to a stack of layers with a single identifier.

### CODE REVIEW

File: build/local_builder.go [hunk 1]
```diff
@@ -135,21 +145,73 @@ func (l *LocalBuilder) PullBaseImageNotExist() (err error) {
...
+		return false
+	}
+
+	baseDir, err := os.Getwd()
+	if err != nil {
+		logger.Error("failed to get current path, err: %v", err)
+		return false
+	}
+
+	layerHash, err := hash.CheckSum(filepath.Join(baseDir, strings.Fields(layer.Value)[0]))
```
Comment (justadogistaken) [inline build/local_builder.go:166]: why checksum under currentDir?

File: build/local_builder.go [hunk 2]
```diff
@@ -135,21 +145,73 @@ func (l *LocalBuilder) PullBaseImageNotExist() (err error) {
...
+	canUseCache := true
 	for i := 1; i < len(l.Image.Spec.Layers); i++ {
 		layer := &l.Image.Spec.Layers[i]
 		logger.Info("run build layer: %s %s", layer.Type, layer.Value)
+		if canUseCache {
+			layers = append(layers, v1.Layer{Value: l.Image.Spec.Layers[i-1].Value, Type: l.Image.Spec.Layers[i-1].Type})
+			img.Spec.Layers = layers
+			config, err := yaml.Marshal(img.Spec.Layers)
+			if err != nil {
+				logger.Error("failed to marshal image into bytes, err: %v", err)
```
Comment (justadogistaken) [inline build/local_builder.go:200]: so, we marshal image or layers above?

File: build/local_builder.go (hunk 1 above)
Comment (justadogistaken) [reply build/local_builder.go:166]: I guess it is a build context, how about put it within builder, get it before building?

File: build/local_builder.go [hunk 3]
```diff
@@ -188,9 +250,29 @@ func (l *LocalBuilder) execCopyLayer(layer *v1.Layer) error {
...
+	if err = l.PlaceSrcHash(layer); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+//This function only has meaning for copy layers
+func (l *LocalBuilder) PlaceSrcHash(layer *v1.Layer) error {
+	baseDir, err := os.Getwd()
```
Comment (justadogistaken) [inline build/local_builder.go:263]: put a context filed within builder

File: image/store/store.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	//read all image layers
+	images := is.Images()
+	for _, image := range images {
+		layers := image.Spec.Layers
+		var tempLayers []v1.Layer
+		for i, layer := range layers {
+			tempLayers = append(tempLayers, v1.Layer{Value: layer.Value, Type: layer.Type})
+			image.Spec.Layers = tempLayers
+			// at present, we only calculate layers without hash
+			storeid, err := GetStoreIDFromImage(*image)
```
Comment (justadogistaken) [inline image/store/store.go:97]: No need to pass whole image into it, does it?

File: image/store/store.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+			if !ok {
+				imagemeta = &imageMeta{
+					layer:    layer,
+					children: make(map[StoreID]struct{}),
+				}
+				is.images[storeid] = imagemeta
+			}
+
+			if i < len(layers)-1 {
+				image.Spec.Layers = append(tempLayers, v1.Layer{Value: layers[i+1].Value, Type: layers[i+1].Type})
```
Comment (justadogistaken) [inline image/store/store.go:113]: will it be nicer to set layers after for done, just use intermediate layer slice?

File: build/local_builder.go [hunk 6]
```diff
@@ -285,9 +367,9 @@ func (l *LocalBuilder) calculateLayerHashAndPlaceIt(layer *v1.Layer, tempTarget
 
 func (l *LocalBuilder) UpdateImageMetadata() error {
 	l.setClusterFileToImage()
-	if err := l.squashBaseImageLayerIntoCurrentImage(); err != nil {
```
Comment (justadogistaken) [inline build/local_builder.go:288]: Canceling squash will make the current image start from "from layer", do you consider the consequence?

**Comment by justadogistaken** [comment]
@touchuyht is that ok for me to continue finishing this pr

**Comment by touchuyht** [comment]
> @touchuyht is that ok for me to continue finishing this pr

Of course. Go on  and finish it.

**Comment by justadogistaken** [comment]
working on the pr #346

**Pull request closed without merging**